bip-0094.mediawiki

```diff
@@ -36,33 +36,54 @@ Since then the issue with block storms has been further demonstrated on Testnet
+In Mainnet and Testnet 3, the difficulty adjustment calculation uses the difficulty value of the last block in the previous period as its base. When the 20-minute exception is applied to this last block, it is mined at difficulty 1, causing the next period's difficulty to be constrained between 1-4, leading to block storms.
 
-=== Time Warp Fix ===
+==== Rule Specification ====
 
-In addition to a time warp attack potentially exacerbating the perpetual block storm attack, a time warp attack provides an alternative way to increase the block production rate even if the unintended reset of the actual difficulty due to the 20-minute exception was mitigated.
+1. For difficulty adjustment calculations between periods:
+   a. The base difficulty value MUST be taken from the first block of the previous difficulty period
+   b. NOT from the last block as in previous implementations
 
-To protect against the time warp attack, the following rule proposed as part of The Great Consensus Cleanup<ref>https://github.com/TheBlueMatt/bips/blob/cleanup-softfork/bip-XXXX.mediawiki</ref> is enforced: "The nTime field of each block whose height, mod 2016, is 0 must be greater than or equal to the nTime field of the immediately prior block minus 600. For the avoidance of doubt, such blocks must still comply with existing Median-Time-Past nTime restrictions."
+2. The adjustment factor calculation remains unchanged:
+   a. Multiplication factor based on the duration of the previous difficulty period
+   b. Limited to no less than 1/4 and no more than 4x
+
+This change ensures that the actual network difficulty is used for adjustment calculations rather than potentially manipulated values from the last block in a period.
+
+=== 3. Time Warp Attack Prevention ===
+
+This rule prevents time warp attacks that could otherwise be used to amplify block storms<ref>A perpetual block storm attack with entire difficulty periods being authored in less than 3.5 days that resets the difficulty to the minimum in the last block of every difficulty period would adjust to a new actual difficulty of 4 every period. An attacker that additionally leverages a time warp attack would start their attack by holding back timestamps until the latest block's timestamp is at least two weeks in the past, and then limiting their block rate to six blocks per second, incrementing the timestamp on every sixth block. Only on the last block they would use the current time, which both resets the difficulty to one per the 20-minute exception and would result in a difficulty adjustment keeping the difficulty at the minimum due to the elapsed time exceeding the target. This would allow lower the difficulty for all blocks to difficulty 1 instead of difficulty 4</ref>.
+
+==== Rule Specification ====
+
+1. For any block whose height modulo 2016 equals 0 (i.e., the first block of each difficulty period):
```

Comment:
I like the "block height modulo 2016 equals 0" definition of "first block". Perhaps this definition could be introduced in advance of Rule 1 as all three rules make reference to the "first block" per this definition.